pkg/alibabacloud/eni/node.go

```diff
@@ -0,0 +1,436 @@
+			continue
+		}
+
+		scopedLog.WithFields(logrus.Fields{
+			fieldENIID:       e.NetworkInterfaceID,
+			"excessIPs":      excessIPs,
+			"freeOnENICount": freeOnENICount,
+		}).Debug("ENI has unused IPs that can be released")
+		maxReleaseOnENI := math.IntMin(freeOnENICount, excessIPs)
+
+		r.InterfaceID = key
+		r.PoolID = ipamTypes.PoolID(e.VPC.VPCID)
+		r.IPsToRelease = freeIpsOnENI[:maxReleaseOnENI]
+	}
+
+	return r
+}
+
+// ReleaseIPs performs the ENI IP release operation
+func (n *Node) ReleaseIPs(ctx context.Context, r *ipam.ReleaseAction) error {
+	return n.manager.api.UnassignPrivateIPAddresses(ctx, r.InterfaceID, r.IPsToRelease)
+}
+
+// GetMaximumAllocatableIPv4 returns the maximum amount of IPv4 addresses
+// that can be allocated to the instance
+func (n *Node) GetMaximumAllocatableIPv4() int {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+
+	// Retrieve l for the instance type
+	l, limitsAvailable := n.getLimitsLocked()
+	if !limitsAvailable {
+		return 0
+	}
+
+	// Return the maximum amount of IP addresses allocatable on the instance
+	// reserve Primary eni
+	return (l.Adapters - 1) * l.IPv4
+}
+
+func (n *Node) loggerLocked() *logrus.Entry {
+	if n == nil || n.instanceID == "" {
+		return log
+	}
+
+	return log.WithField("instanceID", n.instanceID)
+}
+
+// getLimits returns the interface and IP limits of this node
+func (n *Node) getLimits() (ipamTypes.Limits, bool) {
+	n.mutex.RLock()
+	l, b := n.getLimitsLocked()
+	n.mutex.RUnlock()
+	return l, b
+}
+
+// getLimitsLocked is the same function as getLimits, but assumes the n.mutex
+// is read locked.
+func (n *Node) getLimitsLocked() (ipamTypes.Limits, bool) {
+	return limits.Get(n.k8sObj.Spec.AlibabaCloud.InstanceType)
+}
+
+func (n *Node) getSecurityGroupIDs(ctx context.Context, eniSpec eniTypes.Spec) ([]string, error) {
+	// 1. use security group defined by user
+	// 2. use security group used by eth0
+
+	if len(eniSpec.SecurityGroups) > 0 {
+		return eniSpec.SecurityGroups, nil
+	}
+
+	if len(eniSpec.SecurityGroupTags) > 0 {
+		securityGroups := n.manager.FindSecurityGroupByTags(eniSpec.VPCID, eniSpec.SecurityGroupTags)
+		if len(securityGroups) == 0 {
+			n.loggerLocked().WithFields(logrus.Fields{
+				"vpcID": eniSpec.VPCID,
+				"tags":  eniSpec.SecurityGroupTags,
+			}).Warn("No security groups match required vpc id and tags, using eth0 security groups")
+		} else {
+			groups := make([]string, 0, len(securityGroups))
+			for _, secGroup := range securityGroups {
+				groups = append(groups, secGroup.ID)
+			}
+			return groups, nil
+		}
+	}
+
+	var securityGroups []string
+
+	n.manager.ForeachInstance(n.node.InstanceID(),
+		func(instanceID, interfaceID string, rev ipamTypes.InterfaceRevision) error {
+			e, ok := rev.Resource.(*eniTypes.ENI)
+			if ok && e.Type == eniTypes.ENITypePrimary {
+				securityGroups = make([]string, len(e.SecurityGroupIDs))
+				copy(securityGroups, e.SecurityGroupIDs)
+			}
+			return nil
+		})
+
+	if len(securityGroups) <= 0 {
```

Comment:
Is this code supposed to append all the security groups in this foreach iteration? It seems like `securityGroups` will get clobbered in the last iteration of the foreach. Is that right?

Reply:
only have one ENI with type `Primary` , it's ok.

Reply:
If this meant to append all the security groups, then please do as follows:

```suggestion
	var securityGroups []string

	n.manager.ForeachInstance(n.node.InstanceID(),
		func(instanceID, interfaceID string, rev ipamTypes.InterfaceRevision) error {
			e, ok := rev.Resource.(*eniTypes.ENI)
			if ok && e.Type == eniTypes.ENITypePrimary {
				securityGroups = append(securityGroups, e.SecurityGroupIDs...)
			}
			return nil
		})

	if len(securityGroups) <= 0 {
```

Reply:
sure